Use Array.IndexOf to speed up Shroud.Tick.

RoosterDragon opened the pull request:
As the `touched` cell layer uses Boolean values, Array.IndexOf is able to use a fast vectorised search. Most values in the array are false, so the search is able to significantly improve the performance of finding the next true value in the array.

----

Review with ignore whitespace for a cleaner diff.

Running this replay at max speed on commit https://github.com/OpenRA/OpenRA/commit/5157bc375d554e275b142e650fec27356a8bd63b gives the following performance improvement. [ra-2023-11-04T105726Z.zip](https://github.com/OpenRA/OpenRA/files/13256599/ra-2023-11-04T105726Z.zip)

Before
- Shroud.Tick 14.3%
  - OnShroudChanged 2.5%

After
- Shroud.Tick 6.5%
  - UpdateCell 4.0%
      - OnShroudChanged 2.9%
  - IndexOf 1.8%

RoosterDragon committed:
Use Array.IndexOf to speed up Shroud.Tick.

As the `touched` cell layer uses Boolean values, Array.IndexOf is able to use a fast vectorised search. Most values in the array are false, so the search is able to significantly improve the performance of finding the next true value in the array.

RoosterDragon committed:
Use single dictionary call in Shroud.AddSource, Shroud.RemoveSource.

PunkPun reviewed (commented):
Using /all a second time now blinds you. Even when you have explored map option enabled

RoosterDragon committed:
Don't clear/reset shroud when using the `/all` debug command.

Disabling the shroud is sufficient to allow seeing the map. This fixes a game with the "Explored Map" option enabled. Previously using the `/all` command twice to toggle it on and off again would also reset the shroud, causing the map to no longer be explored. Now, using it twice will cause the map to remain explored, as intended when the "Explored Map" option is enabled.

RoosterDragon commented:
> Using /all a second time now blinds you. Even when you have explored map option enabled

Looks like an existing issue on bleed, as using the `/all` command simulates using the "Clear Shroud" and "Reset Shroud" buttons. Have added a commit that removes those buttons from the `/all` command which fixes the problem.

PunkPun commented:
Oh hmm, it looks like this issue in in the current release as well

PunkPun reviewed (approved):
LGTM

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/a6915e441acc1fb282ff7f2f50d223e6bc9326a2...b931e2fde9de5b28a2c6ea4317a4180c5fc69ef2)
